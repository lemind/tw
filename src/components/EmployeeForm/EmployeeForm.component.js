import React from 'react';
import { Field, reduxForm } from 'redux-form';

import { API } from '../../api/index';
import { POSITIONS } from '../../config';

class EmployeeForm extends React.Component {
  constructor(props) {
    super(props);

    this.positionOptions = POSITIONS;

    this.cancelForm = props.cancelForm;
    this.saveForm = props.saveForm;
    this.deleteEmployee = props.deleteEmployee;

    this.submitHandler = this.submitHandler.bind(this);
    this.cancelHandler = this.cancelHandler.bind(this);
    this.deleteHandler = this.deleteHandler.bind(this);
  }

  submitHandler(submittedValues) {
    this.saveForm.apply(null, [submittedValues]);
  }

  deleteHandler(submittedValues) {
    this.deleteEmployee.apply(null, [submittedValues.id]);
  }

  cancelHandler(e) {
    e.preventDefault();
    this.cancelForm.apply();
  }

  render() {
    const { handleSubmit, permission } = this.props;

    return (
      <div>
        <form onSubmit={ handleSubmit(this.submitHandler) }>
          <div className="form-group">
            <label htmlFor="firstName">First name</label>
            <Field
              id="firstName"
              className="form-control"
              name="firstName"
              type="text"
              component="input"
              disabled={ true }
            />
          </div>
          <div className="form-group">
            <label htmlFor="lastName">Last name</label>
            <Field
              id="lastName"
              className="form-control"
              name="lastName"
              type="text"
              component="input"
              disabled={ true }
            />
          </div>
          <div className="form-group">
            <label>Position</label>
            <div>
              <Field
                className="form-control"
                name="position"
                component="select"
                disabled={ !permission }
              >
                <option></option>
                { this.positionOptions.map((n, i) => (
                <option key={ i } value={ n.value }>{ n.label }</option>
                )) }
              </Field>
            </div>
          </div>
          <div className="form-group">
            <label>Role</label>
            <div>
              <label className="radio-inline">
                <Field
                  name="role"
                  component="input"
                  type="radio"
                  value="user"
                  disabled={ !permission }
                />
                User
              </label>
              <label className="radio-inline">
                <Field
                  name="role"
                  component="input"
                  type="radio"
                  value="admin"
                  disabled={ !permission }
                />
                Admin
              </label>
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="experience">Experience</label>
            <Field
              id="experience"
              className="form-control"
              name="experience"
              type="text"
              component="input"
            />
          </div>
          <div className="form-group">
            <label htmlFor="shortDescription">Short description</label>
            <Field
              id="shortDescription"
              className="form-control"
              name="shortDescription"
              type="text"
              component="input"
            />
          </div>
          <div className="form-group">
            <label htmlFor="longDescription">Long description</label>
            <div>
              <Field 
                className="form-control"
                id="longDescription"
                name="longDescription"
                rows="3"
                component="textarea" />
            </div>
          </div>
          <div className="btn-toolbar">
            <button
              type="submit"
              className="btn btn-success"
            >Submit</button>
            <button
              className="btn btn-warning"
              name="cancelButton"
              onClick={ this.cancelHandler }
            >Cancel</button>
            <button
              className="btn btn-danger"
              disabled={ !permission }
              name="deleteButton"
              onClick={ handleSubmit(this.deleteHandler) }
            >Delete</button>
          </div>
        </form>
      </div>
    );
  }
}

export const EmployeeFormComponent = reduxForm({
  form: 'employee'
})(EmployeeForm)

export { EmployeeForm as EmployeeFormPure }