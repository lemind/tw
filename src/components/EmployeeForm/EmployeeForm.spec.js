import React from 'react';
import { shallow, mount } from 'enzyme';
import * as reduxForm from 'redux-form';

import { EmployeeFormPure as EmployeeForm } from './EmployeeForm.component';

class mockField extends React.Component {
  constructor(props) {
    super();
  }

  render() {
    return <div></div>;
  }
}

const handleSubmit = (fn) => {
  return fn;
};

it('should render EmployeeForm', () => {
  expect(shallow(<EmployeeForm handleSubmit={ handleSubmit } />)).toMatchSnapshot();
});

it('should call appropriate function on submit form', () => {
  Object.defineProperty(reduxForm, 'Field', {
    get: () => mockField
  });

  const mockCallback = jest.fn();
  const componentRaw = <EmployeeForm
    handleSubmit={ handleSubmit }
    saveForm={ mockCallback } />;
  const component = mount(componentRaw);
  component.find('[type="submit"]').simulate('submit');
  expect(mockCallback).toBeCalled()
});

it('should call appropriate function on click cancel button', () => {
  Object.defineProperty(reduxForm, 'Field', {
    get: () => mockField
  });

  const mockCallback = jest.fn();
  const componentRaw = <EmployeeForm
    handleSubmit={ handleSubmit }
    cancelForm={ mockCallback } />;
  const component = mount(componentRaw);
  component.find('[name="cancelButton"]').simulate('click');
  expect(mockCallback).toBeCalled()
});

it('should call appropriate function on click delete button if permission is true', () => {
  Object.defineProperty(reduxForm, 'Field', {
    get: () => mockField
  });

  const mockCallback = jest.fn();
  const componentRaw = <EmployeeForm
    handleSubmit={ handleSubmit }
    permission={ true }
    deleteEmployee={ mockCallback } />;
  const component = mount(componentRaw);
  component.find('[name="deleteButton"]').simulate('click');
  expect(mockCallback).toBeCalled()
});

it('should not call appropriate function on click delete button if permission is false', () => {
  Object.defineProperty(reduxForm, 'Field', {
    get: () => mockField
  });

  const mockCallback = jest.fn();
  const componentRaw = <EmployeeForm
    handleSubmit={ handleSubmit }
    permission={ false }
    deleteEmployee={ mockCallback } />;
  const component = mount(componentRaw);
  component.find('[name="deleteButton"]').simulate('click');
  expect(mockCallback).not.toBeCalled()
});
