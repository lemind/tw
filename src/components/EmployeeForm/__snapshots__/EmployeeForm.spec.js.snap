// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EmployeeForm 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="firstName"
      >
        First name
      </label>
      <Field
        className="form-control"
        component="input"
        disabled={true}
        id="firstName"
        name="firstName"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastName"
      >
        Last name
      </label>
      <Field
        className="form-control"
        component="input"
        disabled={true}
        id="lastName"
        name="lastName"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label>
        Position
      </label>
      <div>
        <Field
          className="form-control"
          component="select"
          disabled={true}
          name="position"
        >
          <option />
          <option
            key="0"
            value="admin"
          >
            Admin
          </option>
          <option
            key="1"
            value="developer"
          >
            Developer
          </option>
          <option
            key="2"
            value="manager"
          >
            Manager
          </option>
          <option
            key="3"
            value="tester"
          >
            Tester
          </option>
          <option
            key="4"
            value="hr"
          >
            HR
          </option>
        </Field>
      </div>
    </div>
    <div
      className="form-group"
    >
      <label>
        Role
      </label>
      <div>
        <label
          className="radio-inline"
        >
          <Field
            component="input"
            disabled={true}
            name="role"
            type="radio"
            value="user"
          />
          User
        </label>
        <label
          className="radio-inline"
        >
          <Field
            component="input"
            disabled={true}
            name="role"
            type="radio"
            value="admin"
          />
          Admin
        </label>
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="experience"
      >
        Experience
      </label>
      <Field
        className="form-control"
        component="input"
        id="experience"
        name="experience"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="shortDescription"
      >
        Short description
      </label>
      <Field
        className="form-control"
        component="input"
        id="shortDescription"
        name="shortDescription"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="longDescription"
      >
        Long description
      </label>
      <div>
        <Field
          className="form-control"
          component="textarea"
          id="longDescription"
          name="longDescription"
          rows="3"
        />
      </div>
    </div>
    <div
      className="btn-toolbar"
    >
      <button
        className="btn btn-success"
        type="submit"
      >
        Submit
      </button>
      <button
        className="btn btn-warning"
        name="cancelButton"
        onClick={[Function]}
      >
        Cancel
      </button>
      <button
        className="btn btn-danger"
        disabled={true}
        name="deleteButton"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </form>
</div>
`;
